src/database/post.ts

```diff
@@ -3,92 +3,155 @@ import { PostIdentifier, ProfileIdentifier, Identifier, PostIVIdentifier, GroupI
+import { createDBAccessWithAsyncUpgrade, createTransaction } from './helpers/openDB'
+import type { AESJsonWebKey } from '../modules/CryptoAlgorithm/interfaces/utils'
+import { CryptoKeyToJsonWebKey } from '../utils/type-transform/CryptoKey-JsonWebKey'
 
-const db = createDBAccess(() => {
-    return openDB<PostDB>('maskbook-post-v2', 4, {
-        async upgrade(db, oldVersion, newVersion, transaction): Promise<void> {
-            if (oldVersion < 1) {
-                // inline keys
-                return void db.createObjectStore('post', { keyPath: 'identifier' })
-            }
-            /**
-             * In the version 1 we use PostIdentifier to store post that identified by post iv
-             * After upgrade to version 2, we use PostIVIdentifier to store it.
-             * So we transform all old data into new data.
-             */
-            if (oldVersion <= 1) {
-                const store = transaction.objectStore('post')
-                store.getAll().then((values) => {
-                    store.clear()
-                    for (const each of values) {
+type UpgradeKnowledge = { version: 4; data: Map<string, AESJsonWebKey> } | undefined
+const db = createDBAccessWithAsyncUpgrade<PostDB, UpgradeKnowledge>(
+    4,
+    5,
+    (currentTryOpen, knowledge) =>
+        openDB<PostDB>('maskbook-post-v2', currentTryOpen, {
+            async upgrade(db, oldVersion, _newVersion, transaction): Promise<void> {
+                type Version2PostRecord = {
+                    postBy: PrototypeLess<ProfileIdentifier>
+                    identifier: string
+                    recipientGroups: PrototypeLess<GroupIdentifier>[]
+                    recipients?: ProfileIdentifier[]
+                    foundAt: Date
+                    postCryptoKey?: CryptoKey
+                }
+                type Version3RecipientDetail = {
+                    /** Why they're able to receive this message? */
+                    reason: RecipientReason[]
+                }
+                type Version3PostRecord = Omit<Version2PostRecord, 'recipients'> & {
+                    recipients: Record<string, Version3RecipientDetail>
+                }
+                type Version4PostRecord = Omit<Version3PostRecord, 'recipients'> & {
+                    recipients: Map<string, Version3RecipientDetail>
+                }
+                type Version5PostRecord = Omit<Version4PostRecord, 'postCryptoKey'> & {
+                    postCryptoKey?: AESJsonWebKey
+                }
+                /**
+                 * A type assert that make sure a and b are the same type
+                 * @param a The latest version PostRecord
+                 */
+                function _assert(a: Version5PostRecord, b: PostDBRecord) {
+                    a = b
+                    b = a
+                }
+                // Prevent unused code removal
+                if (1 + 1 === 3) _assert({} as any, {} as any)
```

Comment:
Not understand this.

Reply:
Need to use `_assert` to let typescript check assign-ability for us. A type level hack.

But if `_assert` is not called, it will be removed by the editor because it is unused code